internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,322 @@
+type dynamicScaleDownExecutor interface {
+	ScaleDown() error
+}
+
+// dynamicScaleDownEvaluator is able to perform dynamic ScaleDown evaluation actions
+type dynamicScaleDownEvaluator interface {
+	ShouldScaleDown() (bool, error)
+}
+
+// dynamicScaleDownProcessor is able to process dynamic ScaleDown reconcile events
+type dynamicScaleDownProcessor interface {
+	dynamicScaleDownExecutor
+	dynamicScaleDownEvaluator
+}
+
+// standardDynamicScaleDownProcessor is the default dynamicScaleDownProcessor
+// used when dynamic scaling is enabled.
+// It assumes the provided kafkaStreamingUnitCountPerClusterList does not
+// contain any element with a Status attribute with value 'failed'
+type standardDynamicScaleDownProcessor struct {
+	clusterID                    string
+	regionsSupportedInstanceType config.InstanceTypeMap
+	// kafkaStreamingUnitCountPerClusterList must not contain any element
+	// with a Status attribute with value 'failed'
+	kafkaStreamingUnitCountPerClusterList services.KafkaStreamingUnitCountPerClusterList
+	// indexesOfStreamingUnitForSameClusterID is the index of the straming unit in the "kafkaStreamingUnitCountPerClusterList" that has the
+	// same "clusterID" as the one that's current being processed
+	indexesOfStreamingUnitForSameClusterID []int
+	supportedKafkaInstanceTypesConfig      *config.SupportedKafkaInstanceTypesConfig
+	clusterService                         services.ClusterService
+
+	// dryRun controls whether the ScaleDown method performs real actions.
+	// Useful when you don't want to trigger a real scale down.
+	dryRun bool
+}
+
+var _ dynamicScaleDownEvaluator = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a data plane cluster can de deprovisioned.
+// It returns true if all the following conditions happen:
+// 1. If specified the cluster is empty i.e it does not contain any streaming unit
+// 2. If the cluster can be removed without triggering a scale up action
+// Otherwise false is returned.
+// Note:
+// 1. This method assumes kafkaStreamingUnitCountPerClusterList does not
+// contain elements with the Status attribute with the 'failed' value.
+// Thus, if the type is constructed with the assumptions being true, it
+// can be considered as the 'failed' state Clusters are not included in
+// the calculations.
+// 2. Clusters in deprovisioning and cleanup state are excluded, as clusters into those states don't accept kafka instances anymore.
+// 3. Clusters that are still not ready to accept kafka instance are also excluded from the capacity calculation
+func (p *standardDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
+	// First let's check if the cluster is empty
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		clusterIsNotEmpty := p.kafkaStreamingUnitCountPerClusterList[i].Count > 0
```

Comment:
Is there any situation where **kafkaStreamingUnitCountPerClusterList[i]** can be empty?

Reply:
You mean `.Count` to be `0`?

Yes, if the cluster does not have any Kafka at all.

Reply:
sorry, I updated my comment, `nil`?

Reply:
Thanks.

No, it'll never be `nil`.